### ISSUE TITLE
Regression from MAUI 8: Picker entries blank in release mode (Android)

### ISSUE THREAD

**Issue by BioTurboNick** [dotnet/maui#25935]
### Description

I have a `Picker` that is populated by an array of values. They are mapped to a localized string through a `ValueConverter`.

Since updating to MAUI 9, the picker is now full of blank entries in release mode.

<img src="https://github.com/user-attachments/assets/d8a799e1-2afd-4a6c-bce2-94d57adc1151" width="256"> <img src="https://github.com/user-attachments/assets/a953af5d-e6d6-4c78-abe7-74969758b4cb" width="256">

<img src="https://github.com/user-attachments/assets/77a70b4f-7b1b-4a18-b8eb-c385cf991b76" width="256"> <img src="https://github.com/user-attachments/assets/ca6dbe5a-fd0a-4c5e-a7a1-35bcdb0c1a88" width="256">

### Steps to Reproduce

XAML:
```
<ContentView.Resources>
    <ResourceDictionary>
        <vc:IntToPowerConverter x:Key="intToPower" />
    </ResourceDictionary>
</ContentView.Resources>

<Picker Grid.Column="2"
        FontSize="16"
        VerticalTextAlignment="End"
        ItemDisplayBinding="{Binding Converter={StaticResource intToPower}}">
    <Picker.ItemsSource>
        <x:Array Type="{x:Type x:Int32}">
            <x:Int32>0</x:Int32>
            <x:Int32>1</x:Int32>
            <x:Int32>2</x:Int32>
            <x:Int32>3</x:Int32>
            <x:Int32>4</x:Int32>
            <x:Int32>5</x:Int32>
            <x:Int32>6</x:Int32>
            <x:Int32>7</x:Int32>
            <x:Int32>8</x:Int32>
            <x:Int32>9</x:Int32>
        </x:Array>
    </Picker.ItemsSource>
</Picker>
```

C#:
```
static readonly char[] superscriptCharacters = ['⁰', '¹', '²', '³', '⁴', '⁵', '⁶', '⁷', '⁸', '⁹', '⁻'];
static readonly char[] characters = ['0', '1', '2', '3', '4', '5', '6', '7', '8', '9', '-'];
static readonly Dictionary<char, char> superscriptDictionary = new(superscriptCharacters.Length);
const char times = '×';

public static string GetSuperscript(int exponentValue, CultureInfo culture)
{
    string formatedValue = exponentValue.ToString("D", culture);

    string formattedSuperscripts = string.Empty;

    for (int i = 0; i < formatedValue.Length; i++)
        formattedSuperscripts += superscriptDictionary[formatedValue[i]];

    return formattedSuperscripts;
}

public static string ToPowerOfTen(int value, CultureInfo culture)
{
    string formattedValue = string.Empty;

    if (value == 0) return formattedValue;

    formattedValue += times + 10.ToString(culture);

    if (value == 1) return formattedValue;

    formattedValue += GetSuperscript(value, culture);

    return formattedValue;
}

public class IntToPowerConverter :
    IValueConverter
{
    public object? Convert(object? value, Type targetType, object? parameter, CultureInfo culture)
    {
        if (targetType != typeof(string))
            return new object();

        return value switch
        {
            int intValue => FormatHelpers.ToPowerOfTen(intValue, culture),
            _            => string.Empty
        };
    }

    public object? ConvertBack(object? value, Type targetType, object? parameter, CultureInfo culture) =>
        throw new NotImplementedException();
}

```

### Link to public reproduction project repository

https://github.com/BioTurboNick/MauiPickerItemDisplayBindingBug

### Version with bug

9.0.10 SR1

### Is this a regression from previous behavior?

Yes, this used to work in .NET MAUI

### Last version that worked well

8.0.92 SR9.2

### Affected platforms

Android

### Affected platform versions

Android 14

### Relevant log output

```shell

```

**Comment by mattleibow** [comment]
@simonrozsival  This is using `ItemDisplayBinding` and I know we touched that in .NET 9...

**Comment by BioTurboNick** [comment]
FWIW I have confirmed that if I remove `ItemDisplayBinding=` the items show up.

**Comment by simonrozsival** [comment]
@BioTurboNick I cannot reproduce the problem in .NET 9 in the example you provided with this change to how `superscriptDictionary` is initialized (I assume you're setting this up elsewhere in your original codebase):
```diff
-static readonly Dictionary<char, char> superscriptDictionary = new(superscriptCharacters.Length);
+static readonly Dictionary<char, char> superscriptDictionary = characters.Zip(superscriptCharacters).ToDictionary(x => x.First, x => x.Second);
```

Even without this change, the picker displays two values (`""` and `"x10"`) which is different from the behavior you described.

Could you please share a full sample app which demonstrates the bug? There might be something else that causes the issue outside of the code you shared.

**Comment by BioTurboNick** [comment]
Ugh. I spent some hours trying to pare down my app to find where the bad interaction is occurring, and overshot and lost it.

**Comment by BioTurboNick** [comment]
Fortunately, I'm stubborn.

Reproduction repo: https://github.com/BioTurboNick/MauiPickerItemDisplayBindingBug

**Comment by simonrozsival** [comment]
@BioTurboNick thanks for the repro. I can see the problem, the `ItemDisplayBinding` inherits the data type from its parent scope (`x:DataType="local:MainPageViewModel"`) and compiles the binding with an incompatible type. You need to change your code this way:
```diff
-ItemDisplayBinding="{Binding Converter={StaticResource intToPower}}"
+ItemDisplayBinding="{Binding Converter={StaticResource intToPower}, x:DataType={x:Type x:Int32}}"
```

This should be fixed by https://github.com/dotnet/maui/pull/25173 but I don't know if that was supposed to be released in .NET 9 GA (in which case it didn't work as expected?) or if it will be part of the first servicing release (cc @StephaneDelcroix @PureWeen)

**Comment by BioTurboNick** [comment]
Thanks Simon! Is there a way to make an issue like this noisy during debugging?

**Comment by simonrozsival** [comment]
> Is there a way to make an issue like this noisy during debugging?

This was a bug in Release mode, the app worked correctly in Debug mode, so you weren't supposed to get any warning or binding diagnostics in VS. I opened a PR to fix this: #26359
